src/katsdpcontroller/master_controller.py

```diff
@@ -1013,6 +1029,84 @@ def mass_inform(self, name: str, *args: Any) -> None:
+            port = data['address']['port']
+            return yarl.URL.build(scheme='http', host=ip_addr, port=port)
+
+    async def _update_resource_sensors(self, zk: aiozk.ZKClient) -> None:
+        try:
+            cbf_resources_total = 0
+            cbf_resources_maintenance = 0
+            cbf_resources_free = 0
+            url = await self._mesos_master_url(zk)
+            timeout = aiohttp.ClientTimeout(total=5)
+            draining_machines = set()
+            async with aiohttp.ClientSession(timeout=timeout) as session:
+                async with session.get(url / 'master/maintenance/status') as resp:
+                    data = await resp.json()
+                    draining = data.get('draining_machines', [])
+                    for machine in draining:
+                        draining_machines.add(machine['id']['hostname'])
+                async with session.get(url / 'master/slaves') as resp:
+                    data = await resp.json()
+                    for agent in data['slaves']:
+                        attributes = agent.get('attributes', {})
+                        try:
+                            subsystems_raw = attributes['katsdpcontroller.subsystems']
+                            subsystems = decode_json_base64(subsystems_raw)
+                            SUBSYSTEMS.validate(subsystems)
+                            is_cbf = 'cbf' in subsystems
+                        except Exception:
+                            # Lots of possible exceptions: missing attributes, bad base64,
+                            # bad JSON, schema error...
+                            is_cbf = False
+                        if not is_cbf:
+                            continue
+                        cbf_resources_total += 1
+                        if agent['hostname'] in draining_machines:
+                            cbf_resources_maintenance += 1
+                        elif agent['used_resources']['cpus'] == 0:
+                            cbf_resources_free += 1
+        except Exception as exc:
+            logger.warning('Failed to get resource information from Mesos: %s', exc)
+            self.sensors['cbf-resources-total'].set_value(0, status=Sensor.Status.FAILURE)
+            self.sensors['cbf-resources-maintenance'].set_value(0, status=Sensor.Status.FAILURE)
+            self.sensors['cbf-resources-free'].set_value(0, status=Sensor.Status.FAILURE)
+        else:
+            self.sensors['cbf-resources-total'].value = cbf_resources_total
+            self.sensors['cbf-resources-maintenance'].value = cbf_resources_maintenance
+            self.sensors['cbf-resources-free'].value = cbf_resources_free
+
+    async def _update_resource_sensors_repeat(self) -> None:
+        zk = aiozk.ZKClient(self._args.zk, chroot='mesos', allow_read_only=True)
+        await zk.start()
+        try:
+            while True:
+                await self._update_resource_sensors(zk)
+                await asyncio.sleep(5)
```

Comment:
Not that it's particularly important, and I don't know which behaviour will ultimately be preferable, but this will update every 5 seconds plus however long the update takes.
Alternatively, `gather()`ing these two will bring the update period closer to five seconds properly.
Not that it really matters, I don't think there's a spec, and perhaps ensuring that fully 5 seconds elapse between queries will make things easier for Mesos to breathe.

Reply:
Yeah, I was on the fence, but I figured that if Mesos is somehow overloaded and this process is taking a significant amount of time, then it's probably reasonable to poll less frequently.